src/main/java/com/example/demo/service/InstitutionService.java

```diff
@@ -0,0 +1,142 @@
+package com.example.demo.service;
+
+import com.example.demo.exceptions.*;
+import com.example.demo.model.DoctorRequest;
+import com.example.demo.model.dto.FullInstitutionDTO;
+import com.example.demo.model.dto.InstitutionDTO;
+import com.example.demo.model.dto.PageContentDTO;
+import com.example.demo.model.entity.Doctor;
+import com.example.demo.model.entity.Institution;
+import com.example.demo.model.mapper.InstitutionMapper;
+import com.example.demo.repository.DoctorRepository;
+import com.example.demo.repository.InstitutionRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class InstitutionService {
+    private final InstitutionRepository institutionRepository;
+    private final InstitutionMapper institutionMapper;
+    private final DoctorRepository doctorRepository;
+    private final DoctorService doctorService;
+
+    public PageContentDTO<InstitutionDTO> getInstitutions(Pageable pageable) {
+        Page<Institution> institutionPage = institutionRepository.findAll(pageable);
+        List<InstitutionDTO> institutionDTOS = institutionMapper.map(institutionPage.getContent());
+        return new PageContentDTO<>(
+                institutionPage.getTotalElements(),
+                institutionPage.getNumber(),
+                institutionPage.getTotalPages(),
+                institutionDTOS
+        );
+    }
+
+    public InstitutionDTO getInstitutionByName(String name) {
+        return institutionMapper.map(institutionRepository.findByName(name)
+                .orElseThrow(() -> new InstitutionNotFoundException("Institution with name: " + name + " do not exists")));
+    }
+
+    public void addInstitution(Institution institution) {
+        validateAddingInstitution(institution);
+        institutionRepository.save(institution);
+    }
+
+    public void addInstitutionsWithDoctors(List<FullInstitutionDTO> fullInstitutionDTOS) {
+        Set<Institution> institutions = fullInstitutionDTOS.stream()
+                .map(fullInstitutionDTO -> {
```

Comment:
side effect w streamie raczej unikaj takich rzeczy zastanow sie jak inaczej to napisac